cli/azd/pkg/tools/bicep/bicep.go

```diff
@@ -5,170 +5,237 @@ package bicep
+		); err != nil {
+			return nil, fmt.Errorf("upgrading bicep: %w", err)
+		}
 	}
 
-	return true, nil
+	log.Printf("using local bicep: %s", bicepPath)
+
+	return cli, nil
 }
 
-func (cli *bicepCli) Build(ctx context.Context, file string) (string, error) {
-	args := []string{"build", file, "--stdout"}
-	buildRes, err := cli.runCommand(ctx, args...)
+// runStep runs a long running operation, using the console to show a spinner for progress and status.
+func runStep(ctx context.Context, console input.Console, title string, action func() error) error {
+	console.ShowSpinner(ctx, title, input.Step)
+	err := action()
 
 	if err != nil {
-		return "", fmt.Errorf(
-			"failed running az bicep build: %s (%w)",
-			buildRes.String(),
-			err,
-		)
+		console.StopSpinner(ctx, title, input.StepFailed)
+		return err
 	}
 
-	return buildRes.Stdout, nil
+	console.StopSpinner(ctx, title, input.StepDone)
+	return nil
 }
 
-func (cli *bicepCli) runCommand(ctx context.Context, args ...string) (exec.RunResult, error) {
-	if cli.commandPath == nil {
-		commandPath, err := findBicepPath()
-		if err != nil {
-			return exec.RunResult{}, err
-		}
+type bicepCli struct {
+	path   string
+	runner exec.CommandRunner
+}
+
+// cachedBicepPath returns the path where we cache our local copy of bicep ($AZD_CONFIG_DIR/bin).
+func cachedBicepPath() (string, error) {
+	configDir, err := config.GetUserConfigDir()
+	if err != nil {
+		return "", err
+	}
 
-		cli.commandPath = commandPath
+	if runtime.GOOS == "windows" {
+		return filepath.Join(configDir, "bin", "bicep.exe"), nil
 	}
 
-	runArgs := exec.NewRunArgs(*cli.commandPath, args...)
-	return cli.commandRunner.Run(ctx, runArgs)
+	return filepath.Join(configDir, "bin", "bicep"), nil
 }
 
-const (
-	defaultBicepCommandPath string = "bicep"
-	envNameAzureConfigDir   string = "AZURE_CONFIG_DIR"
-)
+// downloadBicep downloads a given version of bicep from the release site, writing the output to name.
+func downloadBicep(ctx context.Context, transporter policy.Transporter, bicepVersion semver.Version, name string) error {
+	var releaseName string
+	switch runtime.GOOS {
+	case "windows":
+		releaseName = "bicep-win-x64.exe"
+	case "darwin":
+		releaseName = "bicep-osx-x64"
+	case "linux":
+		if _, err := os.Stat("/lib/ld-musl-x86_64.so.1"); err == nil {
+			releaseName = "bicep-linux-musl-x64"
+		} else {
+			releaseName = "bicep-linux-x64"
+		}
+	default:
+		return fmt.Errorf("unsupported platform")
+	}
+
+	bicepReleaseUrl := fmt.Sprintf("https://downloads.bicep.azure.com/v%s/%s", bicepVersion, releaseName)
 
-// Finds the bicep command path
-// Search in PATH, otherwise looks for standalone and az installation locations
-func findBicepPath() (*string, error) {
-	// First, check if tool is found in path
-	// This typically should return true for standalone install and false for `az` bundled install
-	found, err := tools.ToolInPath(defaultBicepCommandPath)
+	log.Printf("downloading bicep release %s -> %s", bicepReleaseUrl, name)
```

Comment:
It might be useful to track this download operation for perf, availability and usage. Would you consider adding an event tracking this using `telemetry.GetTracer().Start(ctx, bicepInstallEvent)`? We should define `bicepInstallEvent` as "tools.bicep.install" in `events.go`

Existing example in `root.go`:
https://github.com/Azure/azure-dev/blob/bdddfae0277e5dd933fc9984444f143333d905ea/cli/azd/cmd/root.go#L176-L184